[backend] Add Gitter backend

animeshk08 opened the pull request:
This commit fixes issue #631 
This commit adds Gitter backend to
Perceval.

Signed-off-by: animesh <animuz111@gmail.com>

animeshk08 commented:
Hi @valeriocos. I have implemented Gitter backend for Perceval. Please try it out using the rooms mentioned here:

Some Gitter instances which can be used for testing may be:

1. https://gitter.im/jenkinsci/jenkins
1. https://gitter.im/fossasia/fossasia
1. https://gitter.im/openMF/mifos

Please note you have to join a room to fetch its messages. Also, an API token is mandatory for this backend. You can Generate one from here: https://developer.gitter.im/docs/welcome,
by Signing in and going to your `profile` or the `Apps` section at the top bar.

Thanks!

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+                            from_archive, self.ssl_verify, room=room)
+
+
+class GitterClient(HttpClient):
```
This class should inherits from [RateLimitHandler](https://github.com/chaoss/grimoirelab-perceval/blob/198c87e34e1acd7c8eee656471d77c812967b505/perceval/client.py#L223) too, since the API responses return rate limit information (see figue below). Such an info should be used to handle requests throttling/consumption.

![captura_277](https://user-images.githubusercontent.com/6515067/77246864-9d7f9880-6c2b-11ea-81e7-161d2b4e1d3c.png)

Example of the use of the RateLimitHandler are:
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/gitlab.py#L464
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/meetup.py#L362
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L571

In a nutshell, you have to:
- check if the ratelimit headers are the default ones (https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/client.py#L235). If not, you should redefine them (ex.: https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/gitlab.py#L464)

- init the instance
  - https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/meetup.py#L362

- redefine calculate_time_to_reset
  - https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/meetup.py#L364
  - https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/gitlab.py#L612
  - https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L625

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+        num_msgs = 0
+        self.room_id = self.client.get_room_id()
+        logger.debug("Room Id of room: %s is %s", self.room, self.room_id)
+        page = 0
```
Please move the variable `page` before the call of `logger.debug`, and separate them with a blank line to improve readibility.

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+        """
+        from_date = kwargs['from_date']
+        logger.debug("Get Gitter message paginated items of room: %s from date: %s", self.room, from_date)
+        from_date = str_to_datetime(from_date).timestamp()
```
Please add a blank line before the assignment of `from_date` to improve readibility. Thus what is after the `logger.debug` statement concerns only the initialization of the variables

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+    :param archive: an archive to store/read fetched data
+    :param from_archive: it tells whether to write/read the archive
+    :param ssl_verify: enable/disable SSL verification
+    :param room: room from which the message are to be fetch.
```
This param may be removed if we decide to not pass the room information to the client.

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+    GET_MESSAGES_ENDPOINT = 'chatMessages'
+    GET_ROOMS_ENDPOINT = 'rooms'
+
+    def __init__(self, api_token, max_items=MAX_ITEMS, archive=None,
```
The client should implement the method `sanitize_for_archive`, which is used to remove the token information from the requests before storing them to the archive (ex.: https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/gitlab.py#L684).

This method isn't present in other clients (like the ones below).
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/pagure.py
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py

 An issue should be opened to check how many clients miss this method and if this is preventing to fetch archived data generated with the corresponding backends. @animeshk08 let me know if you want to open this issue, otherwise I can do it :)

animeshk08 replied:
Thanks for the reference! I will open the issue:)

animeshk08 commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+
+        room = urijoin(self.group, self.room)
+
+        return GitterClient(self.api_token, self.max_items, self.archive,
```
Hi @valeriocos. The only reason I have used roomid to init the client is because, I thought it would be a good idea to include the roomid in the search fields. Since roomid needs a separate API call as is not returned by the chatmessages API.
Let me know if it's not necessary? Or if you prefer any other implementation

valeriocos replied:
Sorry for not being clear @animeshk08 . The roomid in the search fields is a really good idea, but this is at the level of the backend. 

I was suggesting to remove the param `room` at https://github.com/chaoss/grimoirelab-perceval/pull/632/files#diff-6459a75c74db880450cc119cd49e3051R235. The info to retrieve the roomid can be passed as params of the method `get_room_id` (https://github.com/chaoss/grimoirelab-perceval/pull/632/files#diff-6459a75c74db880450cc119cd49e3051R278)

animeshk08 replied:
Thank you for the clarification. I will make the change :)

animeshk08 commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+        This contains the absolute path of the room i.e. {group}/{room}
+    """
+    AUTHORIZATION_HEADER = 'Authorization'
+    GET_MESSAGES_ENDPOINT = 'chatMessages'
```
Sure! Thank you for the info. I will get a PR ready addressing this issue!

valeriocos commented:
@animeshk08 ping me when the PR is ready for review, thanks!

animeshk08 commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,325 @@
+                                              ssl_verify=True)
+
+        # Backend token is required
+        action = parser.parser._option_string_actions['--api-token']
```
Hi @valeriocos. I believe these lines are added to specify that an API token is necessary for this backend.
If these lines are removed and no API token is provided the error throw is: `401 Client Error: Unauthorized for url: https://api.gitter.im/v1/rooms`. 
However, if these are included and no API is provide the error thrown is `error: the following arguments are required: -t/--api-token`. 
Let me know if there is another approach fo making the `api-token` mandatory for this backend. :)

valeriocos replied:
Good catch! I thought that the token was required. Thanks!

animeshk08 commented:
Hi, @valeriocos Sorry for the delay! I have made the changes requested :)

Also, I have added the tests and updated the docs :rocket: 

Can you please review the PR? Thank you :)

valeriocos commented:
Sure @animeshk08 , I'm on it, thanks!

valeriocos commented on tests/test_gitter.py:
```diff
@@ -0,0 +1,498 @@
+    def test_has_resuming(self):
+        """Test if it returns True when has_resuming is called"""
+
+        self.assertEqual(Gitter.has_resuming(), True)
```
This should be changed to `False`

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,378 @@
+
+        :returns: this backend does not support items resuming
+        """
+        return True
```
I have been playing with the backend, and the messages returned are from the most recent to the last recent. In this case, the backend doesn't support resuming, because in case of failure the fetching process must start over.

The resuming is supported when the data is retrieved from the oldest to the most recent, because in case of failures the fetching process can start from the last item collected.

valeriocos replied:
It seems that it isn't possible to fetch the messages in ASC order. If you find something, please let me know.

animeshk08 replied:
Thank you for the info :)
I rechecked the Gitter API doc. The API does not support fetching the messages in ascending order. I will make this change :)

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,378 @@
+    Client for fetching information from the Gitter server
+    using its REST API.
+
+    :param api_token: key needed to use the API
```
sleep_for_rate and min_rate_to_sleep are missing

valeriocos commented on perceval/backends/core/gitter.py:
```diff
@@ -0,0 +1,378 @@
+    def calculate_time_to_reset(self):
+        """Number of seconds to wait. They are contained in the rate limit reset header"""
+
+        time_to_reset = self.rate_limit_reset_ts - (datetime_utcnow().replace(microsecond=0).timestamp() + 1)
```
When the rate limit is over, the time to reset is not logged correctly. Please check if the following code improves the log message:
```
-        time_to_reset = self.rate_limit_reset_ts - (datetime_utcnow().replace(microsecond=0).timestamp() + 1)
+        current_epoch = (datetime_utcnow().replace(microsecond=0).timestamp() + 1) * 1000
+        time_to_reset = (self.rate_limit_reset_ts - current_epoch) / 1000
```

valeriocos commented on tests/test_gitter.py:
```diff
@@ -0,0 +1,498 @@
+                '--min-rate-to-sleep', '1',
+                '--from-date', '2020-03-01',
+                'testapicomm', 'community',
+                ]
```
please move the brakets to the previous line, thanks

animeshk08 commented:
Hi, @valeriocos. I have made the changes. Please review the PR :)
Thanks

animeshk08 commented:
Thanks, @valeriocos. I will start working on the issues and PRs discussed here :)
